utils/crypto.cpp

```diff
@@ -0,0 +1,236 @@
+// as published by the Free Software Foundation.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program. If not, see <http://www.gnu.org/licenses/>.
+
+/**
+ * The uuid/public key/signature validation code for this module was taken
+ * with modification from
+ *
+ * https://eclipsesource.com/blogs/2016/09/07/tutorial-code-signing-and-verification-with-openssl/
+ *
+ * This module duplicates the following functionality that can be performed
+ * on the command line: convert base64 encoded signature to binary
+ *
+ *          openssl base64 -d -in signature.txt -out /tmp/sign.sha256
+ * verify the uuid
+ *          openssl dgst -sha256 -verify public.pem -signature /tmp/sign.sha256 uuid.txt
+ */
+
+#include <include/bluzelle.hpp>
+#include "crypto.hpp"
+#include <openssl/ssl.h>
+#include <openssl/err.h>
+#include <string>
+#include <stdexcept>
+#include <iostream>
+#include <memory>
+
+namespace
+{
+    /**
+     * NOTE This is temporary
+     * We shall use this only until we can get the public key from an Etherium
+     * contract.
+     * TODO Retrieve the public key from the Etherium contract.
+     */
+    const std::string temporary_public_pem
+    {
+            "-----BEGIN PUBLIC KEY-----\n"
+            "MIICIjANBgkqhkiG9w0BAQEFAAOCAg8AMIICCgKCAgEA1KVskYjJo1kMmLPcf1dA\n"
+            "+OxTQD3Rs2ivNOfZe7z8+SZXVEeGXuUAF7qLcza16smpQno2mAL/amYf+BSWuhQf\n"
+            "jCDWVH/ypSg+C0/8Cx0Qksi5Y+Shopq6PJx0aONLUWtuD6bfGLhBylT6GV66aoN4\n"
+            "ozD5mUbPd8/p/5KdLbWTKhCJSqiioa7VmObTegs+CPnCF1KWyLWHSKidZr5ufm2D\n"
+            "wM6IpHIW3ryGCkHSlalV/z2IzSlw+8bVfcQ2+7ZZx+BLmtYd0oxnu2oHPXCKMktK\n"
+            "mKqZV4sSuuOyan7b/nm4nUubCaJYUjoSDRxJ81Ppo94GSc4r8ga4qZShPPHDgSBJ\n"
+            "b3phnOxMdFYcgRTmL7pqobsRiQFJcQgRgX92/vbsZZx+p6axHbT1upBEKxFhn32o\n"
+            "hSUycb6ozTP2vyB5S07wSlla5Grcv6JIVJFbTn5Gw3Je54hMaFVqIKcX4eK5cYZ6\n"
+            "sFrn8c+G/WGC1S8GcWa+7PdMMCSWM9XQjnLoZ8qh80WUfsUTphbdo+ABherxcpeO\n"
+            "P1jN0qg/JNAGAQ9n5BaOoUWBjyN8+oyZ7fnwd4ZGkAJL4XkyGnoFxakVZqtVQGbV\n"
+            "9JBIbFkLtuloHbo0Ax56SDvQ1WDTiusYkkLeU0uUXHcFKa+jcbhjzYqHLW5bA4Mn\n"
+            "g9rJ8ehWRHw3TuXh6EX4xTcCAwEAAQ==\n"
+            "-----END PUBLIC KEY-----"
+    };
+
+
+    /**
+     * This function is used by base_64_decode to determine the amount of memory required
+     * for the binary buffer represented by the base64 encoded string.
+     *
+     * @param base_64_input base64 encoded string
+     * @return a size_t containing the number of bytes expected in the decoded string.
+    */
+    size_t
+    calculate_decode_length(const std::string& base_64_input)
+    {
```

Comment:
What if input is empty?

Reply:
Added a test to the first line of the function that returns 0 if base_64_input is empty.

Also went back up the stack to the base64_decode function to return a failure and throw up a log message if the input that is supposed to contain a base64 encoded string is empty.

Added a unit test as well.